### PR TITLE
[HOPSWORKS-596] Refactor AgentResource

### PR DESCRIPTION
**Pull request by kouzant**
[HOPSWORKS-596] Add agent heartbeat and register views instead of raw String
[HOPSWORKS-596] Refactor agent API [HOPSWORKS-596] Refactor kagent alerting API
[HOPSWORKS-596] Fix licence

[HOPSWORKS-596]  Fix bug with MessageBodyWriter not found for Hosts entity object

[HOPSWORKS-596]  Add Hopsworks CA path to testing environment example

[HOPSWORKS-596]  Add horovod to ProvidedLibraryNames, resolving conflict with HOPSWORKS-532

[HOPSWORKS-596]  Add environment YML attribute to Conda commands view

[HOPSWORKS-596]  Add builders for conda and system commands views

## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [x] Adds tests for the submitted changes (for bug fixes & features)
- [x] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**
https://logicalclocks.atlassian.net/browse/HOPSWORKS-596

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
New registration and CSR signing endpoints for kagent

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:
Same JIRA ticket for
* https://github.com/logicalclocks/kagent-chef/pull/57
* https://github.com/logicalclocks/hopsworks-chef/pull/194
* https://github.com/logicalclocks/conda-chef/pull/10

### CODE REVIEW

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/pythonDeps/PythonDepsFacade.java [hunk 1]
```diff
@@ -82,7 +82,32 @@
 
   private final static Logger logger = Logger.getLogger(PythonDepsFacade.class.
       getName());
-
+  
+  //User upgradable libraries we installed for them
+  public static final ArrayList<String> providedLibraryNames = new ArrayList<String>() {
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/pythonDeps/PythonDepsFacade.java]: @kouzant could we move these to the variables table, so we don't have to redeploy the app to change the list? Unless there is a UI way to modify it (I don't think so but I don't remember for sure).

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java [hunk 2]
```diff
@@ -108,567 +75,146 @@
...
-            pyDep.setVersion(split[1]);
-            deps.add(pyDep);
-          }
-        }
-      }
+  
+  private String handleRegister(AgentView request) throws AppException {
+    logger.log(Level.FINE, "Handling register");
+    if (request == null || request.getHostId() == null || request.getPassword() == null) {
+      logger.log(Level.SEVERE, "Registration request is NULL");
```
[the remark this answers is not in the document]
Comment (kouzant) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java]: Sure, good point

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/pythonDeps/PythonDepsFacade.java (hunk 1 above)
Comment (kouzant) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/pythonDeps/PythonDepsFacade.java]: I'll amend this PR and hopsworks-chef

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java [hunk 3]
```diff
@@ -108,567 +75,151 @@
...
-      }
-    }
-  }
-
-  private void emailAlert(String subject, String body) {
-    try {
-      emailBean.sendEmails(settings.getAlertEmailAddrs(), subject, body);
-    } catch (MessagingException ex) {
-      logger.log(Level.SEVERE, ex.getMessage());
+  public Response handle(AgentView request,
```
Comment (SirOibaf) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java]: You should annotate the endpoint with description and return type for swagger.

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java [hunk 4]
```diff
@@ -108,567 +75,151 @@
...
+        final HeartbeatReplyDTO commands = handleHeartbeat(request);
+        
+        final AgentView agentHBReply = heartbeatReplyToAgentView(commands);
+        
+        return noCacheResponse.getNoCacheResponseBuilder(Response.Status.OK)
+            .entity(agentHBReply).build();
+      case REGISTER:
+        logger.log(Level.FINE, "AgentResource: Register");
+        final String hadoopHome = handleRegister(request);
+        final AgentView agentRegReply = new AgentView();
```
Comment (SirOibaf) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java]: What's the purpose of all these `finals`?

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java [hunk 5]
```diff
@@ -108,567 +75,151 @@
...
-   * @param project
-   * @return
-   */
-  private String listCondaEnvironment(String project) {
-
-    String prog = settings.getHopsworksDomainDir() + "/bin/list_environment.sh";
-    ProcessBuilder pb = new ProcessBuilder(prog, project);
-    StringBuilder sb = new StringBuilder();
+  
+  private void handlePing() {
```
Comment (SirOibaf) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java]: Is this really used?

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/CondaCommandView.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    public Builder setEnvironmentYml(String yml) {
+      this.environmentYml = yml;
+      return this;
+    }
+    
+    public CondaCommandView build() {
+      return new CondaCommandView(this);
+    }
+  
+    public Builder reset() {
```
Comment (SirOibaf) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/CondaCommandView.java]: What's the purpose of this reset method?

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/SystemCommandView.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    public Builder setUser(String user) {
+      this.execUser = user;
+      return this;
+    }
+    
+    public SystemCommandView build() {
+      return new SystemCommandView(this);
+    }
+    
+    public Builder reset() {
```
Comment (SirOibaf) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/SystemCommandView.java]: Same question here.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/agent/AgentController.java [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+            && service.status.equals(Status.Started)) {
+          hostService.setStartTime(heartbeat.agentTime);
+        }
+        hostService.setStatus(service.status);
+      } else {
+        hostService.setStatus(Status.None);
+      }
+      
+      if (service.status.equals(Status.Started)) {
+        hostService.setStopTime(heartbeat.agentTime);
```
Comment (SirOibaf) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/agent/AgentController.java]: did you mean setStartTime? or if status.stopped?

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java [hunk 9]
```diff
@@ -108,567 +75,151 @@
...
-
-  private void emailAlert(String subject, String body) {
-    try {
-      emailBean.sendEmails(settings.getAlertEmailAddrs(), subject, body);
-    } catch (MessagingException ex) {
-      logger.log(Level.SEVERE, ex.getMessage());
+  public Response handle(AgentView request,
+      @QueryParam("action")
+      @DefaultValue("NONE") AgentAction action)
+    throws AppException {
```
[the remark this answers is not in the document]
Comment (kouzant) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java]: That's true, but the general idea of this PR is to required minimum effort to merge to API V2. I suppose @tkakantousis has already implemented Hopsworks exceptions in his branch. So I don't want to implement it twice. When V2 will be merged on master branch I should change the Exception handling

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/CondaCommandView.java (hunk 6 above)
Comment (kouzant) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/CondaCommandView.java]: To reset the Builder to an initial state. When I use the Builder to transform from DAOs to Views I do NOT construct new Builder for each command, instead I used the same builder but reset it.

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (tkakantousis) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java]: That's true, don't do it now.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/agent/AgentController.java (hunk 8 above)
Comment (kouzant) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/agent/AgentController.java]: Nope :) I just move the code from AgentResource to AgentController. But basically the idea is that agent will send a 'Started' status when a service transitioned from 'Stopped' -> 'Started', so 'heartbeat.agentTime' is an approximation of the stop time

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java (hunk 4 above)
Comment (kouzant) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/agent/AgentResource.java]: The heartbeat passes through several layers and I want to be sure it is not re-assigned accidentally, but mainly to emphasize it is and semantically *should* remain immutable

**Pull request merged**